robot/impl/resource_manager.go

```diff
@@ -356,19 +371,36 @@ func (manager *resourceManager) internalResourceNames() []resource.Name {
 func (manager *resourceManager) ResourceNames() []resource.Name {
 	names := []resource.Name{}
 	for _, k := range manager.resources.Names() {
-		if k.API == client.RemoteAPI ||
-			k.API.Type.Namespace == resource.APINamespaceRDKInternal {
-			continue
+		if manager.includeResourceName(k) {
+			names = append(names, k)
 		}
-		gNode, ok := manager.resources.Node(k)
-		if !ok || !gNode.HasResource() {
-			continue
+	}
+	return names
+}
+
+// reachableResourceNames returns the names of all reachable resources in the manager.
+func (manager *resourceManager) reachableResourceNames() []resource.Name {
+	names := []resource.Name{}
+	for _, k := range manager.resources.ReachableNames() {
+		if manager.includeResourceName(k) {
+			names = append(names, k)
 		}
-		names = append(names, k)
 	}
 	return names
 }
 
+func (manager *resourceManager) includeResourceName(k resource.Name) bool {
```

Comment:
My PR initially included an `IsReachable` method but the result was kind of awkward and deadlock-y. Probably because I needed to do your second suggestion of fixing mutexes.

I'd prefer to just change the name to some better if that's okay - I agree with you that `include` is crappy and incomplete. So really this is just a utility function that is meant to apply the same filters that `ResourceNames` applies. So perhaps a lowercase `resourceNames` with a doc-string specifying what exactly is included/excluded is the way to go?

Reply:
Yep

Reply:
done https://github.com/viamrobotics/rdk/pull/4421/commits/2d2fbbaaf3674161afa0745f7ac62f8fb297d4f2